### DISCUSSION TITLE
On the necessity of storing raw-data of embedding attribute

### DISCUSSION THREAD

**Discussion by letian-jiang** [milvus-io/milvus#18363] [Ideas & Feature requests]
As we know, the embeddings of an identical input via different encoders may be quite different. Even if their dimensions match, it is meaningless to apply top-k selection over these embeddings. 

That is to say, if the encoder for embedding is updated, users might have to discard all existing data and re-import the whole collection, in which only the embedding field changed. 

In this scenario, a possible better implementation would be: 
* Milvus storing raw data in object storage
* User creates vector index field by supplying an encoder (or using a default encoder provided by Milvus)
* The encoder could be optimized for high inference throughput

Intuition: the raw data is fixed once it is inserted into Milvus, but its embedding could differ based on the choice of the encoder.

**Comment by yhmo** [comment]
This approach is quite different to milvus original requirement.
It is said "everything can be embedding", there are lot of model/method to convert different things(image/video/autio/text...) into embeddings. Milvus is a database for ANN search for embeddings, not for storing all the original raw content. It is hard to build a perfect service to do everything, store raw content, convert raw data to embeddings, store embeddings and ANN search for embeddings.

We can see there are 3 major area here:
1. store raw content
2. convert raw data to embeddings
3. store embeddings and ANN search for embeddings

In my opinion, user can use database(MySQL/MongoDB/PostgreSQL/SQLite...) that they like to store raw content. We provide a tool for the no.2 area, the [towhee](https://towhee.io/). Milvus is for the no.3 area. So far as I know, some users have begun to use towhee+milvus for their AI product.

**Reply by letian-jiang** [reply]
I agree with your vision. But storing raw blob data does not contradict what "everything can be embedding" is. Embeddings are pointless if they come from different encoders (i.e. latent spaces).

And deploying two databases leads to data redundancy in storage, as well as operational complexity.

**Reply by yhmo** [reply]
> I agree with your vision. But storing raw blob data does not contradict what "everything can be embedding" is. Embeddings are pointless if they come from different encoders (i.e. latent spaces).
> 
> And deploying two databases leads to data redundancy in storage, as well as operational complexity.

Maybe one day we can make it to be a super powerful product, if the market require it.
